conference.js

```diff
@@ -2336,6 +2338,102 @@ export default {
+     * @returns {boolean} - Returns true in order notifications to be dismissed.
+     * @private
+     */
+    _useDevice(device) {
+        switch (device.kind) {
+        case 'videoinput': {
+            APP.store.dispatch(updateSettings({
+                userSelectedCameraDeviceId: device.deviceId
+            }));
+
+            APP.store.dispatch(setVideoInputDevice(device.deviceId));
+            break;
+        }
+        case 'audioinput': {
+            APP.store.dispatch(updateSettings({
+                userSelectedMicDeviceId: device.deviceId
+            }));
+
+            APP.store.dispatch(setAudioInputDevice(device.deviceId));
+            break;
+        }
+        case 'audiooutput': {
+            setAudioOutputDeviceId(
+                device.deviceId,
+                APP.store.dispatch,
+                true)
+            .then(() => logger.log('changed audio output device'))
+            .catch(err => {
+                logger.warn(
+                    'Failed to change audio output device.',
+                    'Default or previously set audio output device will',
+                    ' be used instead.',
+                    err);
+            });
+            break;
+        }
+        }
+
+        return true;
+    },
+
+    /**
+     * Finds a new device by comparing new and old array of devices and dispatches
+     * notification with the new device.
+     * @param {MediaDeviceInfo[]} newDevices - The array of new devices we received.
+     * @param {MediaDeviceInfo[]} oldDevices - The array of the old devices we have.
+     * @private
+     */
+    _checkAndNotifyForNewDevice(newDevices, oldDevices) {
+        // let's find the first device that exists only in new devices
+        const newDevice = newDevices.find(
+            nDevice => !oldDevices.find(
+                device => device.deviceId === nDevice.deviceId));
+
+        if (newDevice) {
+
+            // we want to strip any device details that are not very
+            // user friendly, like usb ids put in brackets at the end
+            let description = newDevice.label;
+            const ix = description.lastIndexOf('(');
+
+            if (ix !== -1) {
+                description = description.substr(0, ix);
+            }
+
+            let deviceTypeKey;
+
+            switch (newDevice.kind) {
+            case 'videoinput': {
+                deviceTypeKey = 'settings.selectCamera';
+                break;
+            }
+            case 'audioinput': {
+                deviceTypeKey = 'settings.selectMic';
+                break;
+            }
+            case 'audiooutput': {
+                deviceTypeKey = 'settings.selectAudioOutput';
+                break;
+            }
+            }
+
+            APP.store.dispatch(showNotification({
+                description,
+                titleKey: 'notify.newDeviceTitle',
+                titleArguments: { 'deviceType': APP.translation.translate(deviceTypeKey).toLowerCase() },
```

Comment:
Is this going to work if we change the language after the notification is displayed?

I vaguely remember there were some i18n tricks to put some special string in the translation text and it will auto translate it. Maybe if something like this exists it will look better in the code. Unfortunately I don't remember...

Reply:
No this will not work, but it is not a big deal as nobody will leave those notifications to stay forever and it is a small text camera, microphone, audio output that will not be translated ... and I'm not sure that generated html elements will be correctly displayed when you insert them in a translation string in a atlaskit notification ....

Reply:
I'm wondering if this is going to do the job - https://www.i18next.com/translation-function/nesting#passing-nesting-to-interpolated . Would you try it?

If we don't have another way to make it work, why don't you have 3 translation keys for every type (only the type will be different in the text)? Isn't that going to fix the problem?

Reply:
This doesn't work, I had tried it. We need to upgrade i18next to make it work.